packages/core-contracts/test/proxy/UniversalProxy.test.js

```diff
@@ -0,0 +1,142 @@
+const { ethers } = hre;
+const assert = require('assert');
+const { assertRevert } = require('@synthetixio/core-js/utils/assertions');
+const { findEvent } = require('@synthetixio/core-js/utils/events');
+
+describe('UniversalProxy', () => {
+  let ForwardingProxy, Instance, Implementation;
+
+  describe('when setting UniversalProxyImplementationMockA as the implementation', () => {
+    before('set up proxy and implementation', async () => {
+      let factory;
+
+      factory = await ethers.getContractFactory('UniversalProxyImplementationMockA');
+      Implementation = await factory.deploy();
+
+      factory = await ethers.getContractFactory('ForwardingProxyMock');
+      ForwardingProxy = await factory.deploy(Implementation.address);
+
+      Instance = await ethers.getContractAt(
+        'UniversalProxyImplementationMockA',
+        ForwardingProxy.address
+      );
+      console.log(`Original implementation Address: ${Implementation.address}`);
+    });
+
+    it('shows that the implementation is set', async () => {
+      assert.equal(await ForwardingProxy.getImplementation(), Implementation.address);
+    });
+
+    describe('when interacting with the implementation via the proxy', async () => {
+      describe('when reading and setting a value that exists in the implementation', () => {
+        before('set a value', async () => {
+          await (await Instance.setA(42)).wait();
+        });
+
+        it('can read the value set', async () => {
+          assert.equal(await Instance.getA(), 42);
+        });
+      });
+
+      describe('when trying to call a function that is not payable', () => {
+        it('reverts', async () => {
+          await assertRevert(
+            Instance.setA(1337, { value: ethers.utils.parseEther('1') }),
+            'non-payable method'
+          );
+        });
+      });
+
+      describe('when reading and setting a value that does not exist in the implementation', () => {
+        let BadInstance;
+
+        before('wrap the implementation', async () => {
+          BadInstance = await ethers.getContractAt(
+            'UniversalProxyImplementationMockB',
+            ForwardingProxy.address
+          );
+        });
+
+        it('reverts', async () => {
+          await assertRevert(BadInstance.getB(), 'function selector was not recognized');
+        });
+      });
+    });
+  });
+
+  describe('when attempts to upgrade to an EOA', () => {
```

Comment:
I think these blocks need to be nested in the previous describe, since they assume that the previous things have happened right?